File: elixir/beer-song/beer.exs
defmodule Beer do

  def verse(0) do
    "No more bottles of beer on the wall, no more bottles of beer.\n" <>
    "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
  end

  def verse(1) do
    "1 bottle of beer on the wall, 1 bottle of beer.\n" <>
    "Take it down and pass it around, no more bottles of beer on the wall.\n"
  end

  def verse(bottles) do
    "#{bottles} bottles of beer on the wall, #{bottles} bottles of beer.\n" <>
    "Take one down and pass it around, " <>
    "#{bottles-1} #{pluralize_bottle(bottles-1)} of beer on the wall.\n"
  end

  defp pluralize_bottle(1), do: "bottle"
  defp pluralize_bottle(_), do: "bottles"

  def sing(from, to // 0) when from >= 0 when to >= 0 do
    Enum.map_join(from..to, "\n", verse(&1)) <> "\n"
  end

end

File: javascript/beer-song/beer.js
module.exports = {
  verse: verse,
  sing: sing
};

function sing(fromBottles, toBottles) {
  toBottles = toBottles || 0;
  if(fromBottles === toBottles) {
    return verse(fromBottles);
  }
  return verse(fromBottles) + "\n" + sing(fromBottles - 1, toBottles);
}

function verse(bottles) {
  var remainingBottles = bottles - 1;
  if (remainingBottles === -1) { remainingBottles = 99; }

  return capitalized(soMany(bottles)) + " of beer on the wall, " +
    soMany(bottles) + " of beer.\n" +
    drinkOrBuyMore(bottles) + ", " +
    soMany(remainingBottles) + " of beer on the wall.\n";
}

function soMany(bottles) {
  if (bottles === 0) { return "no more bottles"; }
  if (bottles === 1) { return "1 bottle"; }
  return bottles + " bottles";
}

function drinkOrBuyMore(bottles) {
  if (bottles === 0) { return "Go to the store and buy some more"; }
  return take(bottles) + " down and pass it around";
}

function take(bottles) {
  if (bottles === 1) { return "Take it"; }
  return "Take one";
}

function capitalized(s) {
  return s[0].toUpperCase() + s.slice(1);
}
